src/master.py

```diff
@@ -68,34 +78,171 @@ def server_access():
     result = local('ssh -o "StrictHostKeyChecking no" %s@%s "exit"' % (config.BOOTSTRAP_USER, public_ip))
     return result.return_code == 0
 
+# TODO: deletion candidate
 @echo_output
 def aws_update_many_projects(pname_list):
     minions = ' or '.join([pname + "-*" for pname in pname_list])
     region = aws.find_region()
     with core.stack_conn(core.find_master(region)):
         sudo("salt -C '%s' state.highstate --retcode-passthrough" % minions)
 
+# TODO: deletion candidate
 @debugtask
 @requires_project
 def aws_update_projects(pname):
     "calls state.highstate on ALL projects matching <projectname>-*"
     return aws_update_many_projects([pname])
 
+@cached
+def _cached_master_ip(master_stackname):
+    "provides a small time saving when remastering many minions"
+    return core.stack_data(master_stackname)[0]['private_ip_address']
+
 @debugtask
 @requires_aws_stack
-def remaster_minion(stackname, master_ip=None):
+def remaster(stackname, new_master_stackname):
     "tell minion who their new master is. deletes any existing master key on minion"
-
-    if not master_ip:
-        newest_master = last(core.find_master_servers(core.active_aws_stacks(core.find_region())))
-        if not newest_master:
-            raise core.NoMasterException("no master servers found")
-        master_ip = core.stack_data(newest_master)[0]['private_ip_address']
-
-    print('re-mastering %s to %s' % (stackname, master_ip))
-
+    # TODO: turn this into a decorator
```

Comment:
Yes, should be turned into something else to avoid leaving an `import` in the wrong scope